### PR TITLE
Peek definition

### LINKED ISSUE #2916: Ctrl+Click Peek at definition does not work
Apparently when the "Peek" option is enabled, we do not open the correct window.

Needs more investigation - I'm not sure this is our fault. Haven't testing on C# on my current machine.

### PR DESCRIPTION
**Pull request by huguesv**
Add support for Peek definition (Alt+F12)
Fixes #2916 Ctrl+Click Peek at definition does not work

I'm still working on a couple of bugs:
- Peek view not using Python editor under a specific scenario
- Editor loses cursor after peek view is promoted to document

Might as well review what's there so far, especially the analyzer changes.

The gist of it is that peek view wants to mark the whole definition in the margin (in blue in screenshots below), and the analyzer was only giving us the location of the identifier. In cases of module or parameter definition, that makes no difference, but for a class or function definition, the identifier is only a subset of the whole definition.

image

image

image

When there are multiple definitions (in C# it happens with partial classes):
image

### CODE REVIEW

File: Python/Product/PythonTools/PythonTools/Intellisense/BufferParser.cs [hunk 1]
```diff
@@ -165,7 +165,8 @@ sealed class BufferParser : IDisposable {
         private void InitBuffer(ITextBuffer buffer, int id = 0) {
             var bi = _services.GetBufferInfo(buffer);
             if (!bi.SetAnalysisEntryId(id)) {
-                Debug.Fail("Buffer is already initialized");
+                // TODO: this can occur when a buffer has multiple views (new window, peek window)
+                //Debug.Fail("Buffer is already initialized");
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Intellisense/BufferParser.cs]: FWIW, I'm already working on fixes for this and the TryDispose below, but they're big (need to reverse event hook relationships all over the place...)

**Comment by huguesv** [comment]
I merged with the #2936 and I got this error when I started a peek session:
```

System.InvalidOperationException: cannot replace existing sink 
at  Microsoft.PythonTools.Editor.PythonTextBufferInfo.AddSink(Object key,  IPythonTextBufferInfoEventSink sink) in  C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Editor\PythonTextBufferInfo.cs:line  201 at  Microsoft.PythonTools.Intellisense.PythonSuggestedActionsSource..ctor(PythonEditorServices  services, ITextView textView, ITextBuffer textBuffer) in  C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\PythonSuggestedActionsSource.cs:line  48 
at  Microsoft.PythonTools.Intellisense.PythonSuggestedActionsSourceProvider.CreateSuggestedActionsSource(ITextView  textView, ITextBuffer textBuffer) in  C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Intellisense\PythonSuggestedActionsSourceProvider.cs:line  41 
at  Microsoft.VisualStudio.Language.Intellisense.Implementation.LightBulbBroker.<>c__DisplayClass50_0.<CreateSourcesForBuffer>b__0(ISuggestedActionsSourceProvider  provider)
at  Microsoft.VisualStudio.Text.Utilities.GuardedOperations.InvokeMatchingFactories[TExtensionInstance,TExtensionFactory,TMetadataView](IEnumerable`1  lazyFactories, Func`2 getter, IContentType dataContentType, Object errorSource)  --- End of stack trace from previous location where exception was thrown --- 
at  Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception  exceptionObject)
--


```

**Comment by huguesv** [comment]
Which also happens when I split an existing window.

**Comment by huguesv** [comment]
```
---&amp;gt; (Inner Exception #0) System.InvalidOperationException: cannot replace existing sink
   at Microsoft.PythonTools.Editor.PythonTextBufferInfo.AddSink(Object key, IPythonTextBufferInfoEventSink sink) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Editor\PythonTextBufferInfo.cs:line 201
   at Microsoft.PythonTools.Navigation.DropDownBarClient..ctor(IServiceProvider serviceProvider, IWpfTextView textView, AnalysisEntry analysisEntry) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\DropDownBarClient.cs:line 81
   at Microsoft.PythonTools.Navigation.CodeWindowManager.AddDropDownBar() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 99
   at Microsoft.PythonTools.Navigation.CodeWindowManager.&amp;lt;AddDropDownBarAsync&amp;gt;d__10.MoveNext() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 70&amp;lt;---
</ExceptionString><DataItems><Data></Data></DataItems><InnerException><ExceptionType>System.InvalidOperationException, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</ExceptionType><Message>cannot replace existing sink</Message><StackTrace>   at Microsoft.PythonTools.Editor.PythonTextBufferInfo.AddSink(Object key, IPythonTextBufferInfoEventSink sink) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Editor\PythonTextBufferInfo.cs:line 201
   at Microsoft.PythonTools.Navigation.DropDownBarClient..ctor(IServiceProvider serviceProvider, IWpfTextView textView, AnalysisEntry analysisEntry) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\DropDownBarClient.cs:line 81
   at Microsoft.PythonTools.Navigation.CodeWindowManager.AddDropDownBar() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 99
   at Microsoft.PythonTools.Navigation.CodeWindowManager.&amp;lt;AddDropDownBarAsync&amp;gt;d__10.MoveNext() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 70</StackTrace><ExceptionString>System.InvalidOperationException: cannot replace existing sink
   at Microsoft.PythonTools.Editor.PythonTextBufferInfo.AddSink(Object key, IPythonTextBufferInfoEventSink sink) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Editor\PythonTextBufferInfo.cs:line 201
   at Microsoft.PythonTools.Navigation.DropDownBarClient..ctor(IServiceProvider serviceProvider, IWpfTextView textView, AnalysisEntry analysisEntry) in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\DropDownBarClient.cs:line 81
   at Microsoft.PythonTools.Navigation.CodeWindowManager.AddDropDownBar() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 99
   at Microsoft.PythonTools.Navigation.CodeWindowManager.&amp;lt;AddDropDownBarAsync&amp;gt;d__10.MoveNext() in C:\Projects\huguesv\PTVS\Python\Product\PythonTools\PythonTools\Navigation\CodeWindowManager.cs:line 70</ExceptionString></InnerException></Exception></TraceRecord>
An unhandled exception of type 'System.AggregateException' occurred in mscorlib.dll
One or more errors occurred.
```

**Comment by zooba** [comment]
They're back on me. Let me finish my IronPython fix and I'll get on it.

**Comment by huguesv** [comment]
Asserts and crashes are gone now :)

The 2 issues listed in the PR description are still there, but they are not critical.

**Comment by zooba** [comment]
:shipit:

**Pull request merged**